### PR TITLE
Added following features that can be directly declared in the config file

### PR DESCRIPTION
**Pull request by slama0077**
🚀 New Configuration Options

1. early_stopping
   - Type: bool
   - Default: False
   - Description: When set to true, training will monitor the loss curve and terminate early once it stops improving.

2. minimum_epochs_before_early_stopping
   - Type: int
   - Default: 5
   - Description: Guarantees that at least this many epochs run before early stopping can kick in. Prevents premature halting on very noisy initial losses.

3. patience_early_stopping
   - Type: int
   - Default: 5
   - Description: Number of epochs with no improvement in training loss after which early stopping triggers. Internally, the implementation still tracks small consecutive improvements to avoid stopping on temporary plateaus.

4. dynamic_learning_rate
   - Type: bool
   - Default: False
   - Description: Enables automatic decay of the learning rate during training based on observed loss trends.

5. factor_dynamic_learning_rate
   - Type: float
   - Default: 0.1
   - Description: When dynamic learning rate is active, multiplies the current learning rate by this factor each time decay is triggered.

6. patience_dynamic_learning_rate
   - Type: int
   - Default: 1
   - Description: Number of epochs with no loss improvement before reducing the learning rate by factor_dynamic_learning_rate.

🎯 Benefits & Value

- More Robust Training
  Early stopping with a configurable “warm‑up” (minimum_epochs_before_early_stopping) and patience window prevents both under‑ and over‑training. Users can let long runs self‑terminate once convergence plateaus, saving compute and guarding against overfitting.

- Flexible Hyperparameter Control
  By separating “minimum epochs” from “patience,” and exposing both as integers, users can finely tune how aggressively the model stops. You no longer need to hack your training script for a two‑stage stopping strategy—everything lives in the config.

- Adaptive Learning Rates for Faster Convergence
  Dynamic LR decay kicks in exactly when the loss plateaus (per patience_dynamic_learning_rate), reducing manual LR-scheduling boilerplate. factor_dynamic_learning_rate gives full control over decay magnitude.

- Out‑of‑the‑Box Best Practices
These features implement widely-adopted training heuristics (early stopping, LR decay) in a plug‑and‑play way—no extra callbacks or custom code required. Newcomers and experts alike get a more reliable, reproducible training loop.

### CODE REVIEW

File: neuralhydrology/training/basetrainer.py [hunk 1]
```diff
@@ -205,11 +212,33 @@ def train_and_validate(self):
...
+        if self._dynamic_learning_rate == True:
+            if self._patience_dynamic_learning_rate == 0:
+                LOGGER.info(f"Dynamic learning rate is set to true, but patience is not set. So, setting patience to default value of 1.")
+                self._patience_dynamic_learning_rate = 1
+            if self._factor_dynamic_learning_rate == 0:
+                LOGGER.info(f"Dynamic learning rate is set to true, but factor is not set. So, setting factor to default value of 0.1.")
+                self._factor_dynamic_learning_rate = 0.1
+            if len(self.cfg.learning_rate.keys()) > 0:
+                LOGGER.info(f"Dynamic learning rate is set to true, so static learning rate is ignored.")  
+            scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(self.optimizer, mode='min', factor=self._factor_dynamic_learning_rate, patience=self._patience_dynamic_learning_rate)
```
Comment (gauchm) [inline neuralhydrology/training/basetrainer.py:234]: can you add a LOGGER.warning() here if `self.cfg.is_continue_training` is True, to note that the scheduler state is reset when resuming training?

File: neuralhydrology/training/basetrainer.py [hunk 2]
```diff
@@ -233,6 +262,15 @@ def train_and_validate(self):
...
                     LOGGER.info(print_msg)
+                
+
+                if self._early_stopping == True:
+                    if epoch > self._minimum_epochs_before_early_stopping:
+                        if (early_stopper.early_stop(valid_metrics['avg_total_loss'])):
+                            LOGGER.info(f"Early stopping triggered at epoch {epoch} with validation loss {valid_metrics['avg_total_loss']:.5f}. Training stopped.")
+                            break
+                if self._dynamic_learning_rate == True:
+                    scheduler.step(avg_losses['avg_total_loss'])
```
Comment (gauchm) [inline neuralhydrology/training/basetrainer.py:273]: The scheduler uses the _training_ loss to determine improvement (but the training loss rarely goes up).
I'd rather use `valid_metrics['avg_total_loss']` here as the metric (like for early stopping).

**Review by gauchm — COMMENTED** [review]
Thanks, those features are great to have (and sorry for the late response)! There are some details of the implementation that will maybe take us a few iterations before merging, but it'll be a nice addition to the library.

Finally, once the changes are addressed we should run an auto-format over the file (especially to respect the 120-char line limit).

**Comment by slama0077** [comment]
Thanks for the comments! I'll fix the things you've mentioned as soon as I can and get back to you.

File: neuralhydrology/training/earlystopper.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            if self.consecutive_improvements < int(round(self.patience / 3)):
+                self.counter += 1
+                print(f"No new min. Counter incremented to {self.counter}")   
+            elif self.consecutive_improvements < int(round(self.patience / 2)):
+                self.counter = 0
+                print(f"Consecutive improvements between 3 and 10. Counter reset to 0")
+            else:
+                self.counter = 0
+                self.min_validation_loss = validation_loss
+                print(f"Since it improved for 10 epochs consecutively, but is still not better than the min_validation, the min validation loss is set to the current loss: {validation_loss}")
```
[the remark this answers is not in the document]
Comment (slama0077) [reply neuralhydrology/training/earlystopper.py:39]: I can change this to the simplest logic. I came up with the idea about the consecutive improvements. Sometimes, the validation loss might plateau very early on during the training and that might not be the "true" loss or at least that might not be the trained parameters we are looking for, because the parameter might still not be properly trained for the training dataset. Or sometimes we might just encounter a noise. That's why I came up with this to check if there's some consecutive improvements even after that plateau, we consider the previous plateau as a noise. I thought this was more robust.

**Comment by gauchm** [comment]
Alright, cool! Before we merge, can I just ask you to send a run with and without early stopping/lr scheduling, just to make sure everything still works as expected? Thanks!
